File: ios/Sources/ILDSDesignSystem/IldsRadio.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_radio.dart + web Radio.tsx. TODO(web/desktop): hover states — mobile-first defer.

public enum IldsRadioSize: Sendable {
    case small, medium, large
}

public struct IldsRadio<Value: Hashable>: View {
    private let value: Value
    @Binding private var selection: Value?
    private let label: String?
    private let size: IldsRadioSize
    private let isDisabled: Bool
    private let hasError: Bool
    private let errorText: String?

    public init(
        value: Value,
        selection: Binding<Value?>,
        label: String? = nil,
        size: IldsRadioSize = .medium,
        isDisabled: Bool = false,
        hasError: Bool = false,
        errorText: String? = nil
    ) {
        self.value = value
        _selection = selection
        self.label = label
        self.size = size
        self.isDisabled = isDisabled
        self.hasError = hasError
        self.errorText = errorText
    }

    private var isSelected: Bool { selection == value }

    public var body: some View {
        VStack(alignment: .leading, spacing: ILDSTokens.sp4) {
            Button {
                guard !isDisabled else { return }
                selection = value
            } label: {
                EmptyView()
            }
            .buttonStyle(IldsRadioButtonStyle(
                isSelected: isSelected,
                label: label,
                size: size,
                isDisabled: isDisabled,
                hasError: hasError
            ))
            .disabled(isDisabled)

            if hasError, let errorText {
                Text(errorText)
                    .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular))
                    .foregroundStyle(ILDSTokens.errorRed600)
            }
        }
        .accessibilityLabel(label ?? String(describing: value))
        .accessibilityAddTraits(isSelected ? .isSelected : [])
    }
}

public struct IldsRadioOption<Value: Hashable>: Identifiable, Sendable where Value: Sendable {
    public let id: Value
    public let value: Value
    public let label: String
    public let isDisabled: Bool

    public init(value: Value, label: String, isDisabled: Bool = false) {
        self.id = value
        self.value = value
        self.label = label
        self.isDisabled = isDisabled
    }
}

public struct IldsRadioGroup<Value: Hashable>: View where Value: Sendable {
    private let options: [IldsRadioOption<Value>]
    @Binding private var selection: Value?
    private let size: IldsRadioSize
    private let isDisabled: Bool
    private let hasError: Bool
    private let errorText: String?
    private let axis: Axis

    public init(
        options: [IldsRadioOption<Value>],
        selection: Binding<Value?>,
        size: IldsRadioSize = .medium,
        isDisabled: Bool = false,
        hasError: Bool = false,
        errorText: String? = nil,
        axis: Axis = .vertical
    ) {
        self.options = options
        _selection = selection
        self.size = size
        self.isDisabled = isDisabled
        self.hasError = hasError
        self.errorText = errorText
        self.axis = axis
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: ILDSTokens.sp8) {
            Group {
                if axis == .vertical {
                    VStack(alignment: .leading, spacing: ILDSTokens.sp8) {
                        radioItems
                    }
                } else {
                    HStack(spacing: ILDSTokens.sp16) {
                        radioItems
                    }
                }
            }
            if hasError, let errorText {
                Text(errorText)
                    .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular))
                    .foregroundStyle(ILDSTokens.errorRed600)
            }
        }
    }

    @ViewBuilder
    private var radioItems: some View {
        ForEach(options) { option in
            IldsRadio(
                value: option.value,
                selection: $selection,
                label: option.label,
                size: size,
                isDisabled: isDisabled || option.isDisabled,
                hasError: hasError
            )
        }
    }
}

private struct IldsRadioIndicator: View {
    let isSelected: Bool
    let size: IldsRadioSize
    let isDisabled: Bool
    let hasError: Bool
    let isPressed: Bool
    @FocusState private var isFocused: Bool

    var body: some View {
        let metrics = IldsRadioMetrics(size: size)
        let colors = IldsRadioColors.resolve(
            isSelected: isSelected,
            isDisabled: isDisabled,
            hasError: hasError,
            isPressed: isPressed,
            isFocused: isFocused
        )

        ZStack {
            Circle()
                .fill(colors.background)
                .frame(width: metrics.outerSize, height: metrics.outerSize)
                .overlay {
                    Circle()
                        .strokeBorder(colors.border, lineWidth: colors.borderWidth)
                }
            Circle()
                .fill(colors.dot)
                .frame(
                    width: isSelected ? metrics.innerSize : 0,
                    height: isSelected ? metrics.innerSize : 0
                )
                .animation(.easeInOut(duration: 0.15), value: isSelected)
        }
        .focusable(!isDisabled)
        .focused($isFocused)
    }
}

private struct IldsRadioButtonStyle: ButtonStyle {
    let isSelected: Bool
    let label: String?
    let size: IldsRadioSize
    let isDisabled: Bool
    let hasError: Bool

    func makeBody(configuration: Configuration) -> some View {
        HStack(alignment: .top, spacing: ILDSTokens.sp8) {
            IldsRadioIndicator(
                isSelected: isSelected,
                size: size,
                isDisabled: isDisabled,
                hasError: hasError,
                isPressed: configuration.isPressed && !isDisabled
            )
            if let label {
                Text(label)
                    .font(IldsRadioMetrics(size: size).labelFont)
                    .foregroundStyle(isDisabled ? ILDSTokens.neutralCoolgray300 : ILDSTokens.neutralCoolgray900)
            }
        }
    }
}

private struct IldsRadioMetrics {
    let outerSize: CGFloat
    let innerSize: CGFloat
    let labelFont: Font

    init(size: IldsRadioSize) {
        switch size {
        case .small:
            outerSize = ILDSTokens.sp16
            innerSize = ILDSTokens.sp8
            labelFont = .system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular)
        case .medium:
            outerSize = ILDSTokens.sp20
            innerSize = 10
            labelFont = .system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightRegular)
        case .large:
            outerSize = ILDSTokens.sp24
            innerSize = ILDSTokens.sp12
            labelFont = .system(size: ILDSTokens.fontSize16, weight: ILDSTokens.fontWeightRegular)
        }
    }
}

private struct IldsRadioColors {
    let background: Color
    let border: Color
    let dot: Color
    let borderWidth: CGFloat

    static func resolve(
        isSelected: Bool,
        isDisabled: Bool,
        hasError: Bool,
        isPressed: Bool,
        isFocused: Bool
    ) -> IldsRadioColors {
        let border: Color
        if isDisabled {
            border = isSelected ? ILDSTokens.neutralCoolgray300 : ILDSTokens.neutralCoolgray200
        } else if hasError {
            border = ILDSTokens.errorRed600
        } else if isPressed {
            border = ILDSTokens.primaryOrange700
        } else if isSelected || isFocused {
            border = ILDSTokens.primaryOrange500
        } else {
            border = ILDSTokens.neutralCoolgray300
        }

        let background: Color
        if isDisabled {
            background = ILDSTokens.neutralCoolgray50
        } else if hasError {
            background = ILDSTokens.errorRed50
        } else if isPressed {
            background = ILDSTokens.primaryOrange100
        } else if isSelected {
            background = ILDSTokens.globalWhite000
        } else {
            background = ILDSTokens.globalWhite000
        }

        let dot: Color
        if isDisabled {
            dot = ILDSTokens.neutralCoolgray300
        } else if hasError {
            dot = ILDSTokens.errorRed600
        } else if isPressed {
            dot = ILDSTokens.primaryOrange700
        } else {
            dot = ILDSTokens.primaryOrange500
        }

        let borderWidth: CGFloat = isSelected || isPressed || isFocused || hasError ? 2 : 1
        return IldsRadioColors(background: background, border: border, dot: dot, borderWidth: borderWidth)
    }
}

File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsRadio.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 13486:38485 — mirrors web/Radio.tsx + lib/ilds_radio.dart.

enum class IldsRadioSize { Small, Medium, Large }

data class IldsRadioOption<T>(
    val value: T,
    val label: String,
    val isDisabled: Boolean = false,
)

@Composable
fun <T> IldsRadio(
    value: T,
    groupValue: T?,
    onValueChange: (T) -> Unit,
    modifier: Modifier = Modifier,
    label: String? = null,
    size: IldsRadioSize = IldsRadioSize.Medium,
    isDisabled: Boolean = false,
    hasError: Boolean = false,
) {
    val metrics = remember(size) { IldsRadioMetrics(size) }
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val isSelected = value == groupValue
    val colors = remember(isSelected, isDisabled, hasError, isPressed) {
        IldsRadioColors.resolve(isSelected, isDisabled, hasError, isPressed)
    }

    Row(
        modifier = modifier.semantics {
            role = Role.RadioButton
            selected = isSelected
        },
        horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp8),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Surface(
            onClick = { if (!isDisabled) onValueChange(value) },
            modifier = Modifier.size(metrics.outerSize),
            enabled = !isDisabled,
            interactionSource = interactionSource,
            shape = CircleShape,
            color = colors.background,
            border = BorderStroke(colors.borderWidth, colors.border),
        ) {
            Box(
                modifier = Modifier.size(metrics.outerSize),
                contentAlignment = Alignment.Center,
            ) {
                if (isSelected) {
                    Surface(
                        modifier = Modifier.size(metrics.dotSize),
                        shape = CircleShape,
                        color = colors.dot,
                    ) {}
                }
            }
        }
        if (label != null) {
            Text(
                text = label,
                style = metrics.labelStyle,
                color = colors.label,
            )
        }
    }
}

@Composable
fun <T> IldsRadioGroup(
    options: List<IldsRadioOption<T>>,
    groupValue: T?,
    onValueChange: (T) -> Unit,
    modifier: Modifier = Modifier,
    size: IldsRadioSize = IldsRadioSize.Medium,
    isDisabled: Boolean = false,
    hasError: Boolean = false,
    errorText: String? = null,
    vertical: Boolean = true,
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
    ) {
        val layoutModifier = if (vertical) Modifier else Modifier
        Column(
            modifier = layoutModifier,
            verticalArrangement = Arrangement.spacedBy(IldsTokens.sp8),
        ) {
            options.forEach { option ->
                IldsRadio(
                    value = option.value,
                    groupValue = groupValue,
                    onValueChange = onValueChange,
                    label = option.label,
                    size = size,
                    isDisabled = isDisabled || option.isDisabled,
                    hasError = hasError,
                )
            }
        }
        if (hasError && errorText != null) {
            Text(
                text = errorText,
                style = TextStyle(
                    fontSize = IldsTokens.fontSize12,
                    lineHeight = 16.sp,
                    fontWeight = IldsTokens.fontWeightRegular,
                ),
                color = IldsTokens.errorRed600,
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

@Immutable
private data class IldsRadioColors(
    val background: Color,
    val border: Color,
    val borderWidth: Dp,
    val dot: Color,
    val label: Color,
) {
    companion object {
        fun resolve(
            isSelected: Boolean,
            isDisabled: Boolean,
            hasError: Boolean,
            isPressed: Boolean,
        ): IldsRadioColors {
            if (isDisabled) {
                return IldsRadioColors(
                    background = IldsTokens.neutralCoolgray50,
                    border = IldsTokens.neutralCoolgray200,
                    borderWidth = 1.dp,
                    dot = IldsTokens.neutralCoolgray300,
                    label = IldsTokens.neutralCoolgray300,
                )
            }
            if (hasError) {
                return IldsRadioColors(
                    background = IldsTokens.globalWhite000,
                    border = IldsTokens.errorRed600,
                    borderWidth = if (isSelected) 2.dp else 2.dp,
                    dot = IldsTokens.errorRed600,
                    label = IldsTokens.neutralCoolgray900,
                )
            }
            val accent = when {
                isPressed && isSelected -> IldsTokens.primaryOrange600
                isSelected -> IldsTokens.primaryOrange500
                isPressed -> IldsTokens.neutralCoolgray400
                else -> IldsTokens.neutralCoolgray600
            }
            return IldsRadioColors(
                background = IldsTokens.globalWhite000,
                border = accent,
                borderWidth = if (isSelected || isPressed) 2.dp else 1.dp,
                dot = if (isSelected) {
                    if (isPressed) IldsTokens.primaryOrange600 else IldsTokens.primaryOrange500
                } else {
                    Color.Transparent
                },
                label = IldsTokens.neutralCoolgray900,
            )
        }
    }
}

@Immutable
private data class IldsRadioMetrics(
    val outerSize: Dp,
    val dotSize: Dp,
    val labelStyle: TextStyle,
) {
    constructor(size: IldsRadioSize) : this(
        outerSize = when (size) {
            IldsRadioSize.Small -> 16.dp
            IldsRadioSize.Medium -> 20.dp
            IldsRadioSize.Large -> 24.dp
        },
        dotSize = when (size) {
            IldsRadioSize.Small -> 8.dp
            IldsRadioSize.Medium -> 10.dp
            IldsRadioSize.Large -> 12.dp
        },
        labelStyle = TextStyle(
            fontSize = when (size) {
                IldsRadioSize.Small -> IldsTokens.fontSize12
                IldsRadioSize.Medium -> IldsTokens.fontSize14
                IldsRadioSize.Large -> IldsTokens.fontSize16
            },
            lineHeight = when (size) {
                IldsRadioSize.Small -> 16.sp
                IldsRadioSize.Medium -> 18.sp
                IldsRadioSize.Large -> 20.sp
            },
            fontWeight = IldsTokens.fontWeightRegular,
        ),
    )
}
